.github/workflows/action.yml

```diff
@@ -6,33 +6,59 @@ on:
+  # See https://docs.github.com/en/actions/using-jobs/using-concurrency and https://docs.github.com/en/actions/learn-github-actions/contexts for more info.
+  group: concurrency-group::${{ github.workflow }}::${{ github.event.pull_request.number > 0 && format('pr-{0}', github.event.pull_request.number) || github.ref_name }}${{ github.ref_name == 'main' && format('::{0}', github.run_id) || ''}}
+  cancel-in-progress: ${{ github.ref_name != 'main' }}
+
 jobs:
 
   build-and-test:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
-    - uses: bazelbuild/setup-bazelisk@v1
-    - run: bazel build --config=built-toolchain //... && bazel test --config=built-toolchain //...
-
-  # TODO: deprecate by Jan 2025 https://bazel.build/release
-  build-e2e-bazel-5:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-    - uses: bazelbuild/setup-bazelisk@v1
-    - run: cd e2e/bazel-5 && bazel build //...
-
-  build-e2e-bazel-6-nobzlmod:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-    - uses: bazelbuild/setup-bazelisk@v1
-    - run: cd e2e/bazel-6 && bazel build //...
+    - uses: actions/checkout@v4
+    - uses: bazel-contrib/setup-bazel@0.8.5
+      with:
+        # Avoid downloading Bazel every time.
+        bazelisk-cache: true
+        # Keep a disk-cache
+        disk-cache: true
+        # Share repository cache between workflows.
+        repository-cache: true
+        # enable some flags for CI
+        bazelrc: |
+          import %workspace%/.aspect/bazelrc/ci.bazelrc
+        # keep a cache for MODULE.bazel repos
+        external-cache: true
+    - run: bazel build //... && bazel test //...
 
-  build-e2e-bazel-7-nobzlmod:
+  e2e-workspace-matrix:
+    strategy:
+      matrix:
+        version:
+          - version: 5.x # TODO: deprecate by Jan 2025 https://bazel.build/release
+            bazelrc:
+          - version: 6.x
+            bazelrc: |
+              import %workspace%/../../.aspect/bazelrc/bazel6.bazelrc
+          - version: 7.x
+            bazelrc: |
+              import %workspace%/../../.aspect/bazelrc/bazel6.bazelrc
+              import %workspace%/../../.aspect/bazelrc/bazel7.bazelrc
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
-    - uses: bazelbuild/setup-bazelisk@v1
-    - run: cd e2e/bazel-7 && bazel build //...
+    - uses: actions/checkout@v4
+    - uses: bazel-contrib/setup-bazel@0.8.5
+      with:
+        # Avoid downloading Bazel every time.
+        bazelisk-cache: true
+        # Keep a disk-cache
+        disk-cache: true
+        # Share repository cache between workflows.
+        repository-cache: true
+        # enable some flags for CI
+        bazelrc: |
+          import %workspace%/../../.aspect/bazelrc/ci.bazelrc
+          ${{ matrix.version.bazelrc }}
+    - run: cd e2e/bazel-workspace && USE_BAZEL_VERSION=${{ matrix.version.version }} bazel build //...
```

Comment:
Is this `bazel` binary actually `bazelisk`?

Reply:
Yes bazel is a symlink to bazelisk setup by the workflow helpers

Reply:
let's make that more explicit across the entire file, I see a point in doing so